### ISSUE TITLE
MoPub ImpressionListener.onImpression notCalled for Native Ads

### ISSUE THREAD

**Issue by UnateD** [mopub/mopub-android-sdk#389]
One line summary of the issue here

- [ x ] I am submitting a bug report for existing functionality
- [ x ] I visited https://developers.mopub.com/ and found no answer
- [ x ] I checked https://twittercommunity.com/c/advertiser-api/mopub to make sure that this issue has not already been filed
- [ x ] I checked to make sure that this issue has not already been filed

#### Steps to reproduce the behavior:
Display ads on custom adapter

#### Expected behavior:
As concisely as possible, describe the expected behavior.

#### Observed behavior:
ImpressionListener.onImpression called

#### Evidence:
No impression data

Good afternoon. To work with mediation, we use both custom adapters and adapters from the mopub network. 
To process impression statistics, we use the method described in this section of the integration instructions: https://developers.mopub.com/publishers/android/impression-data/
Networks are integrated according to the instructions from here: 
https://developers.mopub.com/publishers/android/integrate-networks/

Admob network integrated with mopub mediation tool
Yandex / mtarget networks are integrated as custom connections

**Both providers using for NATIVE ads.**

When using both integration options, the method of retrieving impression information only works in a small number of cases. 
Tell me what affects the method ImpressionListener.onImpression?
Why is it not called for every show under exactly the same conditions? Why does it work equally rarely on a custom connection and on a connection using the admob adapter for a mopub?

Application class

```
@Override
    public void onCreate() {
    ImpressionsEmitter.addListener(listener);
}

private ImpressionListener listener = new ImpressionListener() {
        @Override
        public void onImpression(@NonNull String adUnitId, @Nullable ImpressionData impressionData) {
         //this method called only sometimes for Native Ad, but always for Interstitial ads
        }
    };
```

Part of our Ad manager class:

```
ViewBinder viewBinder = new ViewBinder.Builder(R.layout.adv_mopub_native_layout)
                .mainImageId(R.id.ivMain)
                .iconImageId(R.id.ivIcon)
                .titleId(R.id.tvTitle)
                .textId(R.id.tvText)
                .callToActionId(R.id.btnCallToAction)
                .build();

        MoPubStaticNativeAdRenderer renderer = new MoPubStaticNativeAdRenderer(viewBinder);

        GooglePlayServicesAdRenderer googlePlayServicesAdRenderer = new GooglePlayServicesAdRenderer(
                new MediaViewBinder.Builder(R.layout.adv_admob_native_layout_mopub)
                        .iconImageId(R.id.ivIcon)
                        .titleId(R.id.tvTitle)
                        .textId(R.id.tvText)
                        .callToActionId(R.id.btnCallToAction)
                        .addExtra(GooglePlayServicesAdRenderer.VIEW_BINDER_KEY_PRICE, R.id.tvAdditionalInfo)
                        .addExtra(GooglePlayServicesAdRenderer.VIEW_BINDER_KEY_STAR_RATING, R.id.tvRating)
                        .addExtra(GooglePlayServicesAdRenderer.VIEW_BINDER_KEY_ADVERTISER, R.id.tvSponsored)
                        .privacyInformationIconImageId(R.id.ivPrivacyInformation).build());

        loader.registerAdRenderer(new YandexNativeRenderer(unit.getBlockId())); //Our own class
        loader.registerAdRenderer(googlePlayServicesAdRenderer); //MoPub mediation tool integration
        loader.registerAdRenderer(renderer);
```

Thx

**Comment by dravyan** [comment]
@UnateD, thank you for reporting. Could you confirm the ILRD feature is enabled for your app on the account side? - https://developers.mopub.com/publishers/best-practices/ilrd-guide/#enable-ilrd

Thanks,
MoPub Team

**Comment by UnateD** [comment]
Yes, ILRD feature is enabled and and sometimes we get impression information

**Comment by dravyan** [comment]
@UnateD , could you share more details on the custom connection? Are you saying the `onImpression` callback is not fired for custom networks and is fired correctly for AdMob as a mediated network?

Please note that the network data information is only available for [MoPub supported networks](https://developers.mopub.com/publishers/mediation/supported-mediation-partners/#supported-sdk-networks). For the custom networks, network name [fields listed here](https://developers.mopub.com/publishers/best-practices/ilrd-guide/#ilrd-data) will not be available.

Could you share the steps if you are able to reproduce the issue with [MoPub sample app](https://github.com/mopub/mopub-android-sdk/tree/master/mopub-sample)?

**Comment by UnateD** [comment]
I will prepare an sample app and description of our connection in the near future!
Tell me, please, do I understand correctly - if the provider is not in the list of supported ones (https://developers.mopub.com/publishers/mediation/supported-mediation-partners/#supported-sdk-networks), then ILDR will not receive data on it?
So if we show yandex or mytarget ads through MoPub - will ILRD not be received?

**Comment by dravyan** [comment]
@UnateD , If the network is not part of [supported network list](https://developers.mopub.com/publishers/mediation/supported-mediation-partners/#supported-sdk-networks), the impression data JSON will still be sent but the `network_placement_id` and `network_name` info will not be available for custom networks.
Please share the sample app and steps to reproduce the issue to investigate further.

Thanks,
MoPub Team

**Comment by UnateD** [comment]
Good day! In the process of preparing a sample application, I found a solution to our problem!
It turned out that if you call the following code:
`nativeAd.getBaseNativeAd().setNativeEventListener()`
instead of
`nativeAd.setMoPubNativeEventListener()`

for native ad listener - you will not receive the event `onImpression` , registered in Application class

`ImpressionsEmitter.addListener(listener);`

because the event will fire from the listener 

`BaseNativeAd.NativeEventListener.onAdImpressed` 

where is no `ImpressionData` object

In this case, we needed to receive the event of a click on advertising, so the first listener was used. The second one also solves our problem, but does not intercept the ad Impression event that we expect in the Application class.
I did not find a description of this behavior in the documentation and it seems to me that it would be nice to provide information about this

Thanks so much for helping me find the problem!
